frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+
+			let migrations = T::Migrations::get();
+			for i in 0.. {
+				match Self::exec_migration(&mut meter, &migrations, cursor, i == 0) {
+					None => return meter.consumed,
+					Some(ControlFlow::Break(last_cursor)) => {
+						cursor = last_cursor;
+						break
+					},
+					Some(ControlFlow::Continue(next_cursor)) => {
+						cursor = next_cursor;
+					},
+				}
+			}
+
+			Cursor::<T>::set(Some(MigrationCursor::Active(cursor)));
+
+			meter.consumed
+		}
+	}
+
+	#[pallet::call(weight = T::WeightInfo)]
+	impl<T: Config> Pallet<T> {
+		/// Allows root to set the cursor to any value.
+		///
+		/// Should normally not be needed and is only in place as emergency measure.
+		#[pallet::call_index(0)]
+		pub fn force_set_cursor(
+			origin: OriginFor<T>,
+			cursor: Option<MigrationCursor<T::Cursor, T::BlockNumber>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Cursor::<T>::set(cursor);
+
+			Ok(())
+		}
+
+		/// Clear the entries of the `Historic` set.
+		///
+		/// the `map_cursor` must be set to the last value that was returned by the
+		/// `HistoricCleared` event. The first time `None` can be used. `limit` must be chosen in a
+		/// way to result in a sensible weight.
+		#[pallet::call_index(1)]
+		#[pallet::weight({0})] // FAIL-CI
+		pub fn clear_historic(
+			origin: OriginFor<T>,
+			limit: Option<u32>,
+			map_cursor: Option<Vec<u8>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let next = Historic::<T>::clear(limit.unwrap_or_default(), map_cursor.as_deref());
+			Self::deposit_event(Event::HistoricCleared { next_cursor: next.maybe_cursor });
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Try to make progress on the current migration.
+	///
+	/// Returns whether processing should continue or break for this block. The `meter` contains the
+	/// remaining weight that can be consumed.
+	fn exec_migration(
+		meter: &mut WeightMeter,
+		migrations: &MigrationsOf<T>,
+		mut cursor: ActiveCursorOf<T>,
+		is_first: bool,
+	) -> Option<ControlFlow<ActiveCursorOf<T>, ActiveCursorOf<T>>> {
+		let Some(migration) = migrations.get(cursor.index as usize) else {
+			Self::deposit_event(Event::UpgradeCompleted);
+			Cursor::<T>::kill();
+			T::UpgradeStatusHandler::completed();
+			return None;
+		};
+		if Historic::<T>::contains_key(&migration.id()) {
+			Self::deposit_event(Event::MigrationSkippedHistoric { index: cursor.index });
+			cursor.advance(System::<T>::block_number());
+			return Some(ControlFlow::Continue(cursor))
+		}
+
+		let blocks = System::<T>::block_number().saturating_sub(cursor.started_at);
+		match migration.transactional_step(cursor.inner_cursor.clone(), meter) {
+			Ok(Some(next_cursor)) => {
+				Self::deposit_event(Event::MigrationAdvanced { index: cursor.index, blocks });
+				cursor.inner_cursor = Some(next_cursor);
+
+				// We only do one step per block.
+				if migration.max_steps().map_or(false, |max| blocks > max.into()) {
```

Comment:
should this be >= so blocks does not exceed max? 
```suggestion
				if migration.max_steps().map_or(false, |max| blocks >= max.into()) {
```